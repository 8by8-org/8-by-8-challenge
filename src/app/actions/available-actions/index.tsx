export { AvailableActions } from './available-actions';
