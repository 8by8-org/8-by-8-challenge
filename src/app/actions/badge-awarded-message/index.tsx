export { BadgeAwardedMessage } from './badge-awarded-message';
