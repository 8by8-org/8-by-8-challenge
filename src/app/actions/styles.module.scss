@use '../../styles/partials' as *;

.white_curve {
  display: block;
  width: 100%;
  height: 31px;
  background-color: $color-black-8by8;
  background-image: url('/static/images/pages/actions/white-curve.png');
  background-size: 100% 31px;
  background-position: 0 0;
  background-repeat: no-repeat;
}
