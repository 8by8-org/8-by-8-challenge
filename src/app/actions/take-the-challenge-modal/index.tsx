export { TakeTheChallengeModal } from './take-the-challenge-modal';
