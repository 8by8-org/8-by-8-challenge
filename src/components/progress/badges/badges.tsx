import { BadgeComponent } from './badge';
import type { Badge } from '@/model/types/badge';
import styles from './styles.module.scss';

/**
 * Define Props for Badges component
 */
interface BadgesProps {
  badges: Badge[] | undefined;
}

/**
 * Renders a component to display all badges
 *
 * @param badges - Array of Badges
 *
 * @returns A React component displaying each badge
 *
 * @remarks
 * The array can hold a maximum of 8 badges and pops off badges if the length of the array is greater than 8.
 *
 * The array pushes empty Objects until the length of the array matches 8.
 */
export function Badges({ badges }: BadgesProps): JSX.Element {
  if (badges) {
    while (badges.length > 8) {
      badges.pop();
    }

    while (badges.length < 8) {
      badges.push({});
    }
  }

  return (
    <section className={styles.section_3}>
      {badges?.map((badge, index) => {
        return <BadgeComponent badge={badge} index={index + 1} key={index} />;
      })}
    </section>
  );
}
