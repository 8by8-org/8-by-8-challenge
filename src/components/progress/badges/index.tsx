export { BadgeComponent } from './badge';
export { Badges } from './badges';
