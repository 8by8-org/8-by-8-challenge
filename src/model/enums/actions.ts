export enum Actions {
  ElectionReminders = 'electionReminders',
  RegisterToVote = 'registerToVote',
  SharedChallenge = 'sharedChallenge',
}
