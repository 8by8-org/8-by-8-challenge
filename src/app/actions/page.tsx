'use client';
import { useContextSafely } from '@/hooks/use-context-safely';
import { UserContext } from '@/contexts/user-context';
import { PageContainer } from '@/components/utils/page-container';
import { ConfettiAnimation } from '@/components/utils/confetti-animation';
import { Hero } from './hero/hero';
import { AvailableActions } from './available-actions';
import { hasCompletedAllActions } from './utils/has-completed-all-actions';
import styles from './styles.module.scss';
import { lastContributedToCurrentInviter } from './utils/last-contributed-to-current-inviter';

export default function ActionsPage() {
  const { user, invitedBy } = useContextSafely(UserContext, 'ActionsPage');

  const showBadgeAwardedMessage =
    user &&
    invitedBy &&
    lastContributedToCurrentInviter(user.contributedTo, invitedBy) &&
    !hasCompletedAllActions(user);

  return (
    <PageContainer theme="dark">
      {hasCompletedAllActions(user) && <ConfettiAnimation time={8000} />}
      <Hero />
      <div className={styles.white_curve}></div>
      <section className={styles.actions_section}>
        {showBadgeAwardedMessage && (
          <h3 className={styles.badge_awarded_message}>
            {user!.contributedTo.at(-1)!.challengerName} got a badge!
            <br />
            Here are other actions to help the AAPI community.
          </h3>
        )}
        <AvailableActions />
      </section>
    </PageContainer>
  );
}
