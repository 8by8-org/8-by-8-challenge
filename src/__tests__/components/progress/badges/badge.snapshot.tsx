import '@testing-library/jest-dom';
import { render, cleanup } from '@testing-library/react';
import { BadgeComponent } from '@/components/progress/badges';
import { Actions } from '@/model/enums/actions';

describe('SingleBadgeComponent', () => {
  afterEach(cleanup);
  it('renders default badge unchanged.', () => {
    const { container } = render(<BadgeComponent badge={{}} index={1} />);
    expect(container).toMatchSnapshot();
  });

  it('renders action badge unchanged.', () => {
    const { container } = render(
      <BadgeComponent
        badge={{ action: Actions.VoterRegistration }}
        index={2}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders avatar badge unchanged.', () => {
    const { container } = render(
      <BadgeComponent
        badge={{ playerName: 'Player', playerAvatar: 1 }}
        index={1}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
