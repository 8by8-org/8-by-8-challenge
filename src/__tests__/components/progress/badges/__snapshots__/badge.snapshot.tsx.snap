// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleBadgeComponent renders action badge unchanged. 1`] = `
<div>
  <div
    class="badges"
  >
    <div
      class="blob"
    >
      <img
        alt="player badge background blob"
        class="blob_img blob_img_2"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </div>
    <div
      class="blob_content"
    >
      <img
        alt="action badge icon"
        class="icon"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <h3>
        You Registered
      </h3>
    </div>
  </div>
</div>
`;

exports[`SingleBadgeComponent renders avatar badge unchanged. 1`] = `
<div>
  <div
    class="badges"
  >
    <div
      class="blob"
    >
      <img
        alt="player badge background blob"
        class="blob_img blob_img_1"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </div>
    <div
      class="blob_content"
    >
      <img
        alt="player badge icon"
        class="icon"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <h3>
        Player
      </h3>
    </div>
  </div>
</div>
`;

exports[`SingleBadgeComponent renders default badge unchanged. 1`] = `
<div>
  <div
    class="badges"
  >
    <div
      class="blob"
    >
      <img
        alt="player badge background blob"
        class="disable"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </div>
    <div
      class="blob_content"
    >
      <p
        class="index"
      >
        1
      </p>
    </div>
  </div>
</div>
`;
